### PR TITLE
Godel Plugins Updates

### PR DESCRIPTION
**Pull request by Jmeyer1292**
This PR brings all of the changes I made to the original Godel GUI found in the `godel_plugins` package.

### CODE REVIEW

File: godel_plugins/include/godel_plugins/widgets/robot_blending_widget.h [hunk 1]
```diff
@@ -185,6 +106,10 @@ Q_OBJECT
 	bool call_select_surface_service(godel_msgs::SelectSurface::Request &req);
 	bool call_surface_detection_service(godel_msgs::SurfaceDetection& s);
 	void selected_surface_changed_callback(godel_msgs::SelectedSurfacesChangedConstPtr msg);
+        void select_motion_plan(const std::string& name, bool simulate);
+        void request_available_motions(std::vector<std::string>& plans);
+        void request_load_save_motions(const std::string& path, bool isLoad);
+        void update_motion_plan_list(const std::vector<std::string>& names);
```
Comment (Jmeyer1292) [inline godel_plugins/include/godel_plugins/widgets/robot_blending_widget.h]: There are a lot of indentation changes that have to be fixed because of all the different people with all of the different editors.

File: godel_plugins/src/widgets/robot_blending_widget.cpp [hunk 2]
```diff
@@ -429,244 +491,127 @@ void RobotBlendingWidget::save_robot_scan_parameters()
...
-	t.getBasis().getRPY(rx,ry,rz,1);
-	ui_.LineEditX->setText(QString::number(p.x()));
-	ui_.LineEditY->setText(QString::number(p.y()));
-	ui_.LineEditZ->setText(QString::number(p.z()));
-	ui_.LineEditRx->setText(QString::number(RAD2DEG(rx)));
-	ui_.LineEditRy->setText(QString::number(RAD2DEG(ry)));
-	ui_.LineEditRz->setText(QString::number(RAD2DEG(rz)));
+	QString filepath = QFileDialog::getSaveFileName(this, "Save Motion Plan");
+	// No file selected, return immediately
+	ROS_WARN_STREAM("You want to save motion plan to: " << filepath.toStdString());
```
Comment (Jmeyer1292) [inline godel_plugins/src/widgets/robot_blending_widget.cpp]: WARN -> INFO/DEBUG

File: godel_plugins/src/widgets/robot_blending_widget.cpp [hunk 3]
```diff
@@ -429,244 +491,127 @@ void RobotBlendingWidget::save_robot_scan_parameters()
...
-	ui_.LineEditTrMinAngle->setText(QString::number(RAD2DEG( surface_detection_parameters_.tr_min_angle)));
-	ui_.LineEditTrMaxAngle->setText(QString::number(RAD2DEG( surface_detection_parameters_.tr_max_angle)));
-	ui_.CheckBoxTrNormalConsistency->setChecked(static_cast<bool>(surface_detection_parameters_.tr_normal_consistency));
-
-	ui_.CheckBoxPaEnabled->setChecked(static_cast<bool>(surface_detection_parameters_.pa_enabled));
-	ui_.LineEditPaSegMaxIterations->setText(QString::number(surface_detection_parameters_.pa_seg_max_iterations));
-	ui_.LineEditPaSegDistThreshold->setText(QString::number(surface_detection_parameters_.pa_seg_dist_threshold));
-	ui_.LineEditPaSACPlaneDistance->setText(QString::number(surface_detection_parameters_.pa_sac_plane_distance));
-	ui_.LineEditPaKdtreeRadius->setText(QString::number(surface_detection_parameters_.pa_kdtree_radius));
+	ROS_WARN_STREAM("REQUESTING SAVE!");
```
Comment (Jmeyer1292) [inline godel_plugins/src/widgets/robot_blending_widget.cpp]: WARN -> INFO/DEBUG

File: godel_plugins/src/widgets/scan_tool_configuration_window.cpp [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+  params_.window_width = ui_.LineEditWindowWidth->text().toDouble();
+  params_.min_qa_value = ui_.LineEditMinQAValue->text().toDouble();
+  params_.max_qa_value = ui_.LineEditMaxQAValue->text().toDouble();
+
+}
+
+int godel_plugins::ScanPlanConfigWidget::get_quality_combobox_index()
+{
+
+    for (int i = 0; i != quality_metric_list.size(); i++){
```
Comment (Jmeyer1292) [inline godel_plugins/src/widgets/scan_tool_configuration_window.cpp]: This file, aside from brace location, needs to be updated to the new enumeration based method.

File: godel_plugins/include/godel_plugins/widgets/robot_blending_widget.h [hunk 5]
```diff
@@ -55,93 +57,12 @@ const std::string SURFACE_BLENDING_PARAMETERS_SERVICE = "surface_blending_parame
...
-public:
-
-	godel_msgs::SurfaceDetectionParameters surface_detection_parameters_;
-
-protected:
-
-	Ui::SurfaceDetectionConfigWindow ui_;
-};
-
 class RobotBlendingWidget:  public QWidget
```
Comment (jrgnicho) [inline godel_plugins/include/godel_plugins/widgets/robot_blending_widget.h:66]: Out of curiosity, is the RobotBlendingWidget class taking on the role of the deleted Widged classes that were declared in this file?

File: godel_plugins/include/godel_plugins/widgets/robot_blending_widget.h [hunk 6]
```diff
@@ -185,6 +106,10 @@ Q_OBJECT
 	bool call_select_surface_service(godel_msgs::SelectSurface::Request &req);
 	bool call_surface_detection_service(godel_msgs::SurfaceDetection& s);
 	void selected_surface_changed_callback(godel_msgs::SelectedSurfacesChangedConstPtr msg);
+	void select_motion_plan(const std::string& name, bool simulate);
+	void request_available_motions(std::vector<std::string>& plans);
+	void request_load_save_motions(const std::string& path, bool isLoad);
+	void update_motion_plan_list(const std::vector<std::string>& names);
```
Comment (jrgnicho) [inline godel_plugins/include/godel_plugins/widgets/robot_blending_widget.h:112]: method name formatting has been brought up before in previous PRs, it'll be dealt with in another PR specific to this issue.

File: godel_plugins/include/godel_plugins/widgets/scan_tool_configuration_window.h [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+namespace godel_plugins
+{
+
+class ScanPlanConfigWidget : public ParameterWindowBase
+{
+  Q_OBJECT
+public:
+  ScanPlanConfigWidget(godel_msgs::ScanPlanParameters params);
+  godel_msgs::ScanPlanParameters& params() { return params_; }
+  static QStringList quality_metric_list;
```
Comment (jrgnicho) [inline godel_plugins/include/godel_plugins/widgets/scan_tool_configuration_window.h]: It is my understanding that static class members must be POD as described [here](https://google.github.io/styleguide/cppguide.html#Nonmember,_Static_Member,_and_Global_Functions)

File: godel_plugins/include/godel_plugins/widgets/surface_detection_configuration.h [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+
+namespace godel_plugins
+{
+
+class SurfaceDetectionConfigWidget : public ParameterWindowBase
+{
+  Q_OBJECT
+public:
+  SurfaceDetectionConfigWidget(godel_msgs::SurfaceDetectionParameters params);
```
Comment (jrgnicho) [inline godel_plugins/include/godel_plugins/widgets/surface_detection_configuration.h]: My personal preference is to pass arguments as `const &` whenever possible as it may help avoid the small computational penalty incurred by the extra copying operation.   In this case, that penalty is probably negligible.

File: godel_plugins/src/widgets/robot_blending_plugin.ui [hunk 9]
```diff
@@ -22,790 +22,537 @@
...
-      <string>Simulate</string>
-     </property>
-    </widget>
-   </widget>
-  </widget>
+   </item>
+  </layout>
  </widget>
  <resources/>
  <connections/>
```
Comment (jrgnicho) [inline godel_plugins/src/widgets/robot_blending_plugin.ui:558]: Large chunks of the layout file have been changed, does this equate to major cosmetic changes in the blending gui?

Comment (Jmeyer1292) [inline godel_plugins/src/widgets/robot_blending_plugin.ui:558]: Mostly minor changes:
1. New blend and scan process parameter windows
2. Removed unused tabs for load/save plans
3. Added buttons for simulating trajectories
4. Added options for renaming surfaces

File: godel_plugins/include/godel_plugins/widgets/robot_blending_widget.h (hunk 5 above)
Comment (Jmeyer1292) [inline godel_plugins/include/godel_plugins/widgets/robot_blending_widget.h:66]: Components have been moved around to improve code locality. You'll find the Pose Widget bundled with the macro scan configuration window.

File: godel_plugins/include/godel_plugins/widgets/robot_blending_widget.h (hunk 6 above)
Comment (Jmeyer1292) [inline godel_plugins/include/godel_plugins/widgets/robot_blending_widget.h:112]: ok

File: godel_plugins/include/godel_plugins/widgets/scan_tool_configuration_window.h (hunk 7 above)
Comment (Jmeyer1292) [inline godel_plugins/include/godel_plugins/widgets/scan_tool_configuration_window.h]: I've never been particularly fond of this code. We need a QStringList to populate a drop-down menu for surface roughness metric. I can hide it in the .cpp code though.

File: godel_plugins/src/widgets/robot_blending_widget.cpp [hunk 10]
```diff
@@ -53,11 +69,17 @@ void RobotBlendingWidget::init()
...
+	robot_blend_config_window_= new BlendingPlanConfigWidget(blending_plan_parameters_);
+	scan_plan_config_window_=new ScanPlanConfigWidget(scan_plan_parameters_);
 
 	// setting signals and slots
 	connect(robot_scan_config_window_,SIGNAL(parameters_changed()),this,SLOT(robot_scan_params_changed_handler()));
 	connect(surface_detect_config_window_,SIGNAL(parameters_changed()),this,SLOT(surface_detect_params_changed_handler()));
 	connect(ui_.PushButtonMoreOptions,SIGNAL(clicked()),this,SLOT(scan_options_click_handler()));
+
+	connect(ui_.PushButtonBlendOptions,SIGNAL(clicked()),this,SLOT(blend_options_click_handler()));
+	connect(ui_.pushButtonProfileOptions,SIGNAL(clicked()),this,SLOT(scan_plan_options_click_handler()));
```
Comment (jrgnicho) [inline godel_plugins/src/widgets/robot_blending_widget.cpp:81]: Perhaps we should decide on a naming convention for these widget elements in the qt layout as lacking such a convention will lead to inconsistent naming practices.  I propose that we use the camel case convention , beginning with an uppercase and where the last word is indicative of the widget type, as follows : `ui_.FixEverythingMagicalButton`

File: godel_plugins/include/godel_plugins/widgets/surface_detection_configuration.h (hunk 8 above)
Comment (Jmeyer1292) [inline godel_plugins/include/godel_plugins/widgets/surface_detection_configuration.h]: +1
Just an oversight.

File: godel_plugins/src/widgets/robot_blending_widget.cpp [hunk 11]
```diff
@@ -429,244 +491,126 @@ void RobotBlendingWidget::save_robot_scan_parameters()
...
 
-SurfaceDetectionConfigWidget::SurfaceDetectionConfigWidget(godel_msgs::SurfaceDetectionParameters params)
+void RobotBlendingWidget::request_load_save_motions(const std::string& path, bool isLoad)
 {
-	surface_detection_parameters_ = params;
-	init();
-	update_parameters();
-}
+	// Pre-condition: The path must not be an empty string
+	if (path.empty()) return;
```
Comment (jrgnicho) [inline godel_plugins/src/widgets/robot_blending_widget.cpp]: Should this at least throw a warning?

File: godel_plugins/src/widgets/robot_blending_widget.cpp [hunk 12]
```diff
@@ -429,244 +491,126 @@ void RobotBlendingWidget::save_robot_scan_parameters()
...
+	godel_msgs::LoadSaveMotionPlan srv;
+	srv.request.path = path;
 
-void SurfaceDetectionConfigWidget::init()
-{
-	ui_.setupUi(this);
+	if (isLoad)
+		srv.request.mode = godel_msgs::LoadSaveMotionPlan::Request::MODE_LOAD;
+	else
+		srv.request.mode = godel_msgs::LoadSaveMotionPlan::Request::MODE_SAVE;
```
Comment (jrgnicho) [inline godel_plugins/src/widgets/robot_blending_widget.cpp:572]: As this isn't necessarily a one line _if statement_ it may merit braces.

File: godel_plugins/src/widgets/robot_blending_widget.cpp (hunk 11 above)
Comment (Jmeyer1292) [inline godel_plugins/src/widgets/robot_blending_widget.cpp]: I think that the file selection box used to identify the file to opened or closed will return an empty string if the user cancels it. The check is put off till this point.

File: godel_plugins/src/widgets/robot_blending_widget.cpp [hunk 13]
```diff
@@ -429,244 +491,126 @@ void RobotBlendingWidget::save_robot_scan_parameters()
...
-	hide();
+	if (!new_text.isEmpty())
+	{
+		godel_msgs::RenameSurfaceResponse res;
+		godel_msgs::RenameSurfaceRequest req;
+		req.old_name = old_text.toStdString();
+		req.new_name = new_text.toStdString();
+		if (rename_surface_client_.call(req, res))
+		{
+			item->setText(new_text);
```
Comment (jrgnicho) [inline godel_plugins/src/widgets/robot_blending_widget.cpp:619]: Indenting appears to be using more than two spaces

File: godel_plugins/src/widgets/robot_scan_configuration.cpp [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+  // create transform
+  tf::Vector3 p(x,y,z);
+  tf::Quaternion q;
+  q.setRPY(rx,ry,rz);
+
+  return tf::Transform(q,p);
+}
+
+// Scan Config Widget
+godel_plugins::RobotScanConfigWidget::RobotScanConfigWidget(godel_msgs::RobotScanParameters params)
```
Comment (jrgnicho) [inline godel_plugins/src/widgets/robot_scan_configuration.cpp]: Const ref ?

File: godel_plugins/src/widgets/scan_tool_configuration_window.cpp [hunk 15]
```diff
@@ -0,0 +1,69 @@
+#include "godel_plugins/widgets/scan_tool_configuration_window.h"
+
+godel_plugins::ScanPlanConfigWidget::ScanPlanConfigWidget(godel_msgs::ScanPlanParameters params)
```
Comment (jrgnicho) [inline godel_plugins/src/widgets/scan_tool_configuration_window.cpp]: Const ref

File: godel_plugins/src/widgets/robot_blending_widget.cpp (hunk 12 above)
Comment (Jmeyer1292) [inline godel_plugins/src/widgets/robot_blending_widget.cpp:572]: I don't follow, are you saying this is okay?

File: godel_plugins/src/widgets/scan_tool_configuration_window.cpp [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+}
+
+int godel_plugins::ScanPlanConfigWidget::get_quality_combobox_index()
+{
+  switch (params_.quality_metric)
+  {
+  case godel_msgs::ScanPlanParameters::METHOD_RMS:
+    return 0;
+  default:
+    return -1;
```
Comment (jrgnicho) [inline godel_plugins/src/widgets/scan_tool_configuration_window.cpp:57]: I assume that more `ScanPlanParameter` options will be added eventually.

File: godel_plugins/src/widgets/robot_blending_widget.cpp (hunk 11 above)
Comment (Jmeyer1292) [inline godel_plugins/src/widgets/robot_blending_widget.cpp]: I changed this to check the string when the user returns from the file selection menu. There is now a warning here as this function should not get empty paths.

File: godel_plugins/include/godel_plugins/widgets/robot_blending_widget.h (hunk 5 above)
Comment (jrgnicho) [inline godel_plugins/include/godel_plugins/widgets/robot_blending_widget.h:66]: +1

File: godel_plugins/include/godel_plugins/widgets/scan_tool_configuration_window.h (hunk 7 above)
Comment (jrgnicho) [inline godel_plugins/include/godel_plugins/widgets/scan_tool_configuration_window.h]: Sounds good to me

File: godel_plugins/src/widgets/robot_blending_widget.cpp (hunk 13 above)
Comment (Jmeyer1292) [inline godel_plugins/src/widgets/robot_blending_widget.cpp:619]: The whole file is tab spaced, because the original was tab spaced.

File: godel_plugins/src/widgets/robot_scan_configuration.cpp (hunk 14 above)
Comment (Jmeyer1292) [inline godel_plugins/src/widgets/robot_scan_configuration.cpp]: ok

File: godel_plugins/src/widgets/scan_tool_configuration_window.cpp (hunk 15 above)
Comment (Jmeyer1292) [inline godel_plugins/src/widgets/scan_tool_configuration_window.cpp]: ok

File: godel_plugins/src/widgets/robot_blending_plugin.ui (hunk 9 above)
Comment (jrgnicho) [inline godel_plugins/src/widgets/robot_blending_plugin.ui:558]: Ok, thanks for clarifying

File: godel_plugins/src/widgets/scan_tool_configuration_window.cpp (hunk 16 above)
Comment (Jmeyer1292) [inline godel_plugins/src/widgets/scan_tool_configuration_window.cpp:57]: Yes

File: godel_plugins/src/widgets/robot_blending_widget.cpp (hunk 11 above)
Comment (jrgnicho) [inline godel_plugins/src/widgets/robot_blending_widget.cpp]: +1

File: godel_plugins/src/widgets/robot_blending_widget.cpp (hunk 10 above)
Comment (Jmeyer1292) [inline godel_plugins/src/widgets/robot_blending_widget.cpp:81]: We tried to keep it fairly consistent. The only discrepancy is that some start with capital letters and others do not. For ease of integration, I would like to open an issue to make these names consistent. I'll need to change the forms to match.

File: godel_plugins/src/widgets/robot_blending_widget.cpp (hunk 12 above)
Comment (jrgnicho) [inline godel_plugins/src/widgets/robot_blending_widget.cpp:572]: I think this needs braces as it is not a _one line_ if statement

File: godel_plugins/src/widgets/robot_blending_widget.cpp (hunk 13 above)
Comment (jrgnicho) [inline godel_plugins/src/widgets/robot_blending_widget.cpp:619]: Could it be reformatted?

File: godel_plugins/src/widgets/robot_blending_widget.cpp (hunk 10 above)
Comment (jrgnicho) [inline godel_plugins/src/widgets/robot_blending_widget.cpp:81]: +1

File: godel_plugins/src/widgets/robot_blending_widget.cpp (hunk 13 above)
Comment (Jmeyer1292) [inline godel_plugins/src/widgets/robot_blending_widget.cpp:619]: Yes. For the sake of these pull requests, I have:
- Made all new files follow ROS conventions (2-wide spaces)
- Kept original files tabbed.

Because most of the original Godel code was tab spaced, changing it all in the pull requests would read as an absolute sea of changes. It's very easy to change with editors, though, so it's a prime candidate for an issue/immediate pull request after.

Comment (jrgnicho) [inline godel_plugins/src/widgets/robot_blending_widget.cpp:619]: That's a very valid point, let's address this separately.

**Comment by jrgnicho** [comment]
This looks good to me. +1 you can merge
